### PR TITLE
Allow reading config values with a namespace to any value with a ConfigReader

### LINKED ISSUE #263: Unexpected behavior with Option
I'm seeing some inconsistent and unexpected behavior with `Option`.

When a field is optional in a config class, it defaults to `None`, so this works:

```
import com.typesafe.config.ConfigFactory
import pureconfig._

val emptyConfig = ConfigFactory.parseString("")

case class A(a: Option[Int])
assert(A(None) == loadConfigOrThrow[A](emptyConfig))
```

If, however, you move the `Option` into the parameters for `loadConfigOrThrow`, no such defaulting occurs, and the load fails with an exception:

```
case class B(a: Int)
assert(None == loadConfigOrThrow[Option[B]](emptyConfig, "b")) 
// fails with exception
```

Also, defaults are not respected:

```
case class C(a: Option[Int] = Some(10))
assert(C(Some(10)) == loadConfigOrThrow[C](emptyConfig))
// fails, a == None
```

For case `B`, it appears a top-level check would work, something like: 
```
def handleOption[T: ConfigReader : ClassTag](
  config: Config, 
  path: String
): Either[ConfigReaderFailures, T] = {
  val reader = implicitly[ConfigReader[T]]
  val isOption = classOf[Option[_]] == implicitly[ClassTag[T]].runtimeClass
  reader.from(
    if(isOption && !config.hasPath(path)) null else config.getValue(path)
  )
}
```

### PR DESCRIPTION
**Pull request by ruippeixotog**
Replaces the algorithm for loading configs with namespaces as described in #263. It also fixes some bugs:
- `loadConfig` no longer throws exceptions when the namespace does not represent a config object;
- there were some cases where `improveFailures` was not being called after a `ConfigReader` was used. Now all code paths pass through `improveFailures`.

~~Since we want to navigate through the config one hop at a time, I needed a way to parse the string namespace into a proper path. The process is not straightforward, as Typesafe Config accepts special chars and escape sequences in their paths. It employs internally a [parser](https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Path.java) with a full fledged [tokenizer](https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/Tokenizer.java), both of which are unfortunately package private. After deliberating a bit, I decided that it would be worth to use the object anyway, as I prefer to rely on their internal code (the library is not updated that much) than to replicate their logic or stop being fully compatible with their paths. Let me know if you disagree with this, any suggestions are welcome!~~ (thanks for finding a solution to this, @jcazevedo) 

Closes #263.

### CODE REVIEW

**Review by jcazevedo — CHANGES_REQUESTED** [review]
Looks good to me! The improved test suite should help clarify the behaviour of `loadConfig` with a namespace.

I think you can use [`ConfigUtil`](https://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigUtil.html) to access the internal parser for namespaces, thus avoiding direct usage of package private code.

**Comment by ruippeixotog** [comment]
@jcazevedo I didn't know about `ConfigUtil`, great! I changed the wrapper to use it in f8e0a6f (the wrapper is still useful because the Typesafe Config parser doesn't deal with empty lists and path expressions, which I conveniently use to denote the root).

**Review by jcazevedo — APPROVED** [review]
Looks good to me. Thanks for this!

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,11 +1,15 @@
 ### 0.7.3 (unreleased)
 
 - New features
+  - `loadConfig` methods now allow loading any type from a config when using a namespace, and not only types represented
+    by config objects;
   - `ConfigFieldMapping` now has a `withOverrides` method that allows users to easily define exceptional cases to an
     existing mapping;
-  - `ConfigReader` and `ConfigWriter` for `java.math.BigDecimal`, `java.math.BigInteger`, 
+  - `ConfigReader` and `ConfigWriter` for `java.math.BigDecimal` and `java.math.BigInteger`.
```
Comment (melrief) [inline CHANGELOG.md:8]: You are good in writing changelogs 👍

File: core/src/main/scala/pureconfig/package.scala [hunk 2]
```diff
@@ -7,35 +7,63 @@
...
+            case childCv =>
+              getValue(childCv, remaining, key :: curr)
+          }
+
+        case _ => fail(WrongType(
+          cv.valueType, Set(ConfigValueType.OBJECT), ConfigValueLocation(cv.origin()), joinPath(curr.reverse)))
+      }
+    }
+
+    // we're not expecting any exception here, this `try` is just for extra safety
```
Comment (melrief) [inline core/src/main/scala/pureconfig/package.scala:43]: Good idea

**Review by melrief — APPROVED** [review]
Great PR, in particular the tests are a semi-specification of the different errors that we expect. Thanks for doing this.

**Pull request merged**